### ISSUE TITLE
predictCells values

### ISSUE THREAD

**Issue by kmshort** [statOmics/tradeSeq#151]
Hi @HectorRDB, @koenvandenberge  et. al.
I'm using the code from your workshop as a basis to generate a matrix of scaled values from predictCells and plot them on a heatmap (using ComplexHeatmap, if it matters).
 I use ComnplexHeatmap::Heatmap because I can grab the row order from one heatmap, and apply it to another heatmap (which is great because I can then split the heatmap by conditions and maintain pseudotime ordering, while having matched rows).
However, I am sad to say that I don't know what those predictCells values actually *mean* in relation to actual gene expression of those genes, represented as rows, across pseudotime.
Can you please explain it, as you would explain it to a 2 year old?

**Comment by HectorRDB** [comment]
No worries, that function is a bit weird. 
For each cell, we have a pseudotime, and the count value for each gene. However, that count expression is noisy, i.e., the same cell with the same pseudotime values could have had different values for the counts of the genes. 

With the fitGam function, we estimate a model for gene expression based on pseudotime. The predictCells function then output the average count value for all cells and the specified genes. 
For heatmap however, we recommand the predictSmooth function, which will produce estimates of average gene counts for equaly spaced values of pseudotimes. It should produce smoother plots.

Let me know if this is unclear.
Best

**Comment by kmshort** [comment]
Many thanks @HectorRDB 
But I think you mean "However, that count expression is noisy, i.e., *different cells* with the same pseudotime values could have had different values for the counts of the genes". fitGam smooths over the differences, sort of doing what amounts to the equivalent of an advanced "median" filter in image processing.

The clincher in your explanation for me understanding this is when you said "equally spaced values of psesudotime" for the predictSmooth function. On a heatmap you have a continuous scale from 0 to whatever, and gene expression plotted over that scale. But with predictCells when plotted across the heatmap the best you can do is plot them as a sorted order based on their discrete pseudotimes. So what you get is an order of expression in relation to each other, not along pseudotime as such.

The two functions would produce the same result If the cells were perfectly evenly spaced across pseudotime. Which in reality never really happens.

So, what are the scaled units? As in, the units transformed in the section of the bioc2000trajectories workshop:
https://kstreet13.github.io/bioc2020trajectories/articles/workshopTrajectories.html
at "Visualization of DE genes"? I'm using that as a guide.

**Comment by HectorRDB** [comment]
Thanks @kmshort 
I was indeed incorrect, it's the average value for all cells with those pseudotime values and all other characteristics, such as library size or batch effect.
 
"The two functions would produce the same result If the cells were perfectly evenly spaced across pseudotime." So this is not exactly true. `predictSmooth` would return the same values as `predictCells` if the cells were perfectly evenly spaced across pseudotime AND all cells had the same library size.

For the scaled units, as in the code, we scale the smoothed expression gene by gene such that the mean is zero and the variance is 1. This allows for better visualisation in heatmaps.

**Comment by kmshort** [comment]
Okay, thanks again @HectorRDB I didn't realise that it also took into account library size, etc.

I have a question about how you applied predictSmooth in your workshop, where you have two conditions. 

```
### based on mean smoother
yhatSmooth <- predictSmooth(sce, gene = conditionGenes, nPoints = 50, tidy = FALSE)
yhatSmoothScaled <- t(scale(t(yhatSmooth)))
heatSmooth_TGF <- pheatmap(yhatSmoothScaled[, 51:100],
  cluster_cols = FALSE,
  show_rownames = FALSE, show_colnames = FALSE, main = "TGF-Beta", legend = FALSE,
  silent = TRUE
)

matchingHeatmap_mock <- pheatmap(yhatSmoothScaled[heatSmooth_TGF$tree_row$order, 1:50],
  cluster_cols = FALSE, cluster_rows = FALSE,
  show_rownames = FALSE, show_colnames = FALSE, main = "Mock",
  legend = FALSE, silent = TRUE
)

grid.arrange(heatSmooth_TGF[[4]], matchingHeatmap_mock[[4]], ncol = 2)
```

From my interpretation of this, predictSmooth must use the conditions in the tradeSeq object, and calculate smoothers for each condition independently? Since you're using predictSmooth and scaling the matrix from that, I assume that it must. Otherwise, you'd be generating smoothers for both TGF and Mock separately, and then scaling each.

Continuing on from this, if predictCells estimates expression on a discrete per-cell basis, presumably we split the data into conditions before scaling expression matrices? I've tried scaling separately and together. Scaling separately looks good, together is a mess.

**Comment by HectorRDB** [comment]
To explain the code in words:
L1: generate the smooth expression value for the genes. The results is a matrix with one row per gene, and 100 columns (nPoints * 2 conditions)
L2: scale the gene expression. In the first version of the workshop, we scaled the conditions separately but we realized this would not work properly and create artificial differences. 
L3-7: generate heatmap for the TGF condition
L9-13: generate heatmap for the mock condition 
L15: plot the two together

I don't know what your use case is, but if the goal is to generate heatmaps, I would really advice to use the `predictSmooth` function instead of `predictCells`. We also have less experience with that tool so it's harder to diagnose
